lib/shim/shim.js

```diff
@@ -1009,35 +1011,59 @@ function bindCallbackSegment(args, cbIdx, parentSegment) {
     cbIdx = normalizedCBIdx
   }
 
-  // Pull out the callback and make sure it is a function.
+  // Make sure cb is function before wrapping
+  if (this.isFunction(args[cbIdx])) {
+    wrapCallback({ shim: this, args, cbIdx, parentSegment, spec })
+  }
+}
+
+/**
+ * Wraps the callback and creates a segment for the callback function.
+ * It will also call an after hook with the arguments passed to callback
+ *
+ * @private
+ * @param {Object} params to function
+ * @param {Shim} params.shim instance of shim
+ * @param {Array | object} params.args
+ *  The arguments array to pull the cb from.
+ * @param {number|string} params.cbIdx
+ *  The index of the callback.
+ * @param {TraceSegment} [params.parentSegment]
+ *  The segment to use as the callback segment's parent. Defaults to the
+ *  currently active segment.
+ * @param {Spec} params.spec spec to original wrapped function, used to call after method with arguments passed to callback
+ *
+ */
+function wrapCallback({ shim, args, cbIdx, parentSegment, spec }) {
   const cb = args[cbIdx]
-  if (this.isFunction(cb)) {
-    const shim = this
-    const realParent = parentSegment || shim.getSegment()
-    args[cbIdx] = shim.wrap(cb, null, function callbackWrapper(shim, fn, name) {
-      return function wrappedCallback() {
-        if (realParent) {
-          realParent.opaque = false
-        }
-        const segment = _rawCreateSegment(
-          shim,
-          new specs.SegmentSpec({
-            name: 'Callback: ' + name,
-            parent: realParent
-          })
-        )
-
-        if (segment) {
-          segment.async = false
-        }
+  const realParent = parentSegment || shim.getSegment()
+  args[cbIdx] = shim.wrap(cb, null, function callbackWrapper(shim, fn, name) {
+    return function wrappedCallback() {
+      if (realParent) {
+        realParent.opaque = false
+      }
+      const segment = _rawCreateSegment(
+        shim,
+        new specs.SegmentSpec({
+          name: 'Callback: ' + name,
+          parent: realParent
+        })
+      )
 
-        // CB may end the transaction so update the parent's time preemptively.
-        realParent && realParent.touch()
-        return shim.applySegment(cb, segment, true, this, arguments)
+      if (segment) {
+        segment.async = false
       }
-    })
-    shim.storeSegment(args[cbIdx], realParent)
-  }
+
+      if (spec?.after) {
+        spec.after({ shim, fn, name, args: arguments, segment: realParent })
```

Comment:
We need to update the docs at: https://github.com/newrelic/node-newrelic/blob/fcb285a11acca890b90c01dce87ae9fc9c4225fd/lib/shim/specs/recorder.js#L22-L36

In short, the parameters in the doc block need to be shifted into an object.